pkg/datastore/headers.go

```diff
@@ -0,0 +1,433 @@
+	ds "github.com/ipfs/go-datastore"
+)
+
+const (
+	HeaderPrefix        = "header-"
+	HeaderContentPrefix = "hc/"
+
+	InMemorySlotLag        = 200
+	InMemorySlotPurgeCheck = time.Duration(time.Second * 10)
+	InMemorySlotTimeLag    = time.Duration(time.Minute * 1)
+)
+
+var (
+	HeaderPrefixBytes = []byte("header-")
+)
+
+func HeaderKeyContent(slot structs.Slot, blockHash string) ds.Key {
+	return ds.NewKey(fmt.Sprintf("hc/%d/%s", slot, blockHash))
+}
+
+func HeaderMaxNewKey(slot structs.Slot) ds.Key {
+	return ds.NewKey(fmt.Sprintf("hm/%d", slot))
+}
+
+func HeaderKey(slot uint64) ds.Key {
+	return ds.NewKey(fmt.Sprintf("%s%d", HeaderPrefix, slot))
+}
+
+// OLD Entries
+func HeaderMaxProfitKey(slot structs.Slot) ds.Key {
+	return ds.NewKey(fmt.Sprintf("header/max-profit/%d", slot))
+}
+
+type StoredIndex struct {
+	Index                []IndexEl
+	MaxProfit            IndexEl
+	SubmissionsByPubKeys map[[48]byte]IndexEl
+}
+
+func NewStoreIndex() StoredIndex {
+	return StoredIndex{
+		SubmissionsByPubKeys: make(map[[48]byte]IndexEl),
+	}
+}
+
+type IndexEl struct {
+	Hash          [32]byte
+	Value         *big.Int
+	BuilderPubkey [48]byte
+}
+
+type Info struct {
+	Slot  uint64
+	Added time.Time
+}
+
+func (s *Datastore) GetMaxProfitHeader(ctx context.Context, slot structs.Slot) (structs.HeaderAndTrace, error) {
+	// Check memory
+	p, ok := s.hc.GetMaxProfit(uint64(slot))
+	if ok {
+		return p, nil
+	}
+
+	// Check new
+	p, err := s.getMaxHeader(ctx, slot)
+	if err != nil {
+		return p, nil
+	}
+
+	// Check old (until ttl passes)
+	headers, err := s.deprecatedGetHeaders(ctx, HeaderMaxProfitKey(slot))
+	if err != nil {
+		return p, nil
+	}
+
+	if len(headers) == 0 {
+		return p, fmt.Errorf("there is no header")
+	}
+
+	return headers[0], nil
+}
+
+var ErrNotFound = errors.New("not found")
+
+func (s *Datastore) getMaxHeader(ctx context.Context, slot structs.Slot) (h structs.HeaderAndTrace, err error) {
+	txn := s.Badger.NewTransaction(false)
+	defer txn.Discard()
+	defer txn.Commit()
+
+	item, err := txn.Get(HeaderMaxNewKey(slot).Bytes())
+	if err != nil {
+		return h, err
+	}
+
+	item, err = txn.Get(HeaderKeyContent(slot, item.String()).Bytes())
+	if err != nil {
+		return h, err
+	}
+
+	h = structs.HeaderAndTrace{}
+	err = item.Value(func(val []byte) error {
+		return json.Unmarshal(val, &h)
+	})
+
+	return h, err
+}
+
+func (s *Datastore) PutHeader(ctx context.Context, hr structs.HR, ttl time.Duration) (err error) {
+	if err := storeHeader(s.Badger, hr, ttl); err != nil {
+		return err
+	}
+
+	newlyCreated, err := s.hc.Add(uint64(hr.Slot), hr.HeaderAndTrace)
+	if err != nil {
+		return err
+	}
+
+	if !newlyCreated {
+		return // success
+	}
+
+	// check and load keys if exists
+	return s.loadKeysAndCleanup(ctx, uint64(hr.Slot))
+}
+
+func (s *Datastore) loadKeysAndCleanup(ctx context.Context, slot uint64) error {
+	s.l.Lock()
+	defer s.l.Unlock()
+
+	// need to load keys to memory
+	data, err := s.TTLStorage.Get(ctx, HeaderKey(slot))
```

Comment:
Is there an easy way to avoid doing a `Get` call to the DB in every `PutHeader`?

Reply:
It's called only upon the very first payload to the new slot.
It's needed in the scenario when someone would like to add data into the slot that's not in memory. So it loads it on demand (if needed)

Reply:
Aaah I missed the newlyCreated conditional on `PutHeader`. Nice 👍